Added new security blog post (#38758)
supabase/supabase: apps/www/_blog/2025-09-16-defense-in-depth-mcp.mdx

---

> title: Defense in Depth for AI Development, Lessons from the Supabase MCP Debate

Author: I was going to go with: Defense in Depth for MCP Servers

Thoughts? Too reductive?

description: Learn about the security risks of connecting AI agents to databases and how to implement defense in depth strategies to protect your data from prompt injection attacks.
author: bilharmer
image: 2025-09-defense-in-depth-mcp/og.png
thumb: 2025-09-defense-in-depth-mcp/og.png
categories:
  - postgres
  - mcp
  - security
date: 2025-09-16:08:00
toc_depth: 2
---

> Over the past couple of months, there’s been renewed discussion around the risks of connecting AI agents to databases containing private data. A recent blog post by the team at General Analysis ran the headline “Supabase MCP can leak your entire SQL database.” They went on to show that if you spin up a Supabase instance with Row Level Security and a default MCP server accessed through Cursor you could create a scenario where a Stored Prompt Injection attack could be launched. They put instructions into data fields that would direct the MCP server to pull private data from the database and write it back to the text field the attacker was able to see when a developer used an AI agent to connect and read those fields. (full post [here](https://www.generalanalysis.com/blog/supabase-mcp-blog)).

Reviewer: ```suggestion
Over the past few months, there’s been renewed discussion around the risks of connecting MCP servers to databases containing private data. A recent blog post by the team at General Analysis ran the headline “Supabase MCP can leak your entire SQL database.” They went on to show that if you spin up a Supabase instance with Row Level Security and a default MCP server accessed through Cursor you could create a scenario where a Stored Prompt Injection attack could be launched. They put instructions into data fields that would direct the MCP server to pull private data from the database and write it back to the text field the attacker was able to see when a developer used an AI agent to connect and read those fields. (full post [here](https://www.generalanalysis.com/blog/supabase-mcp-blog)).
```

> My initial reaction was to pick apart the inadequacies of this test but I realized that this is the new reality. Vibe coders are not creating separate Production and Staging environments, they are developing realtime on production databases.

Reviewer: ```suggestion
My initial reaction was to debate the MCP server setup but I realized that this is the new reality. Vibe coders are not creating separate Production and Staging environments, they are developing on production databases.
```

> [**Simon Wilson**](https://x.com/simonw) first raised the issue in June of this year with his blog post on [**The Lethal Trifecta**](https://simonwillison.net/2025/Jun/16/the-lethal-trifecta/). He described this as bringing Access to Private Data together with the Ability to Externally Communicate and Exposure to Untrusted Content or in technical terms:

Reviewer: ```suggestion
[Simon Willison](https://x.com/simonw) first raised the issue in June of this year with his blog post on [The Lethal Trifecta](https://simonwillison.net/2025/Jun/16/the-lethal-trifecta/). He described this as bringing Access to Private Data together with the Ability to Externally Communicate and Exposure to Untrusted Content or in technical terms:
```

1. An **LLM** capable of interpreting natural language instructions
2. **Autonomous tool calling** (functions, APIs, or MCP tools)
3. **Access to private or sensitive data sources**

> When these three elements combine without strong controls, data exposure becomes possible, no matter whose MCP server you’re using. This configuration would never pass the go/no go assessment of a Security team…..but there are no security teams in vibe coding. Someone with a great idea is building and deploying to the world solo. It’s up to us to help them deploy as securely as possible, so that’s why I’m actually thankful this issue has been raised.

Reviewer: ```suggestion
When these three elements combine without strong controls, data exposure becomes possible, no matter whose MCP server you’re using. This configuration would never pass the go/no-go assessment of a Security team…but there are no security teams in vibe coding. Someone with a great idea is building and deploying to the world solo. It’s up to us to help them deploy as securely as possible.
```

This problem applies to **any** tool, API, or database connection where a large language model (LLM) can make iterative calls to retrieve or manipulate data. There have been documented issues with [GitHub’s MCP](https://invariantlabs.ai/blog/mcp-github-vulnerability), [Claude Desktop](https://joedivita.substack.com/p/mcp-its-the-wild-west-out-there), and [WhatsApp MCP](https://invariantlabs.ai/blog/whatsapp-mcp-exploited) as well.

> It’s worth clarifying a misconception in the post: **Supabase has never offered a hosted MCP server** (though it is on our roadmap)**.** Our MCP implementation is open source and designed for developers to self-host or be hosted by a 3rd party (Cursor, Cline, etc).

Reviewer: ```suggestion
It’s worth clarifying a misconception in the post: **Supabase has never offered a hosted MCP server** (though it is on our roadmap). Our MCP implementation is open source and designed for developers to self-host or be hosted by a 3rd party (Cursor, Cline, etc).
```

> though it is on our roadmap

should  we say more about how we are going to keep this secure?
